broker/src/main/java/org/apache/rocketmq/broker/pop/PopConsumerService.java

```diff
@@ -0,0 +1,702 @@
+        return brokerController.getConsumerOrderInfoManager().checkBlock(
+            context.getAttemptId(), topicId, groupId, queueId, context.getInvisibleTime());
+    }
+
+    protected CompletableFuture<PopConsumerContext> getMessageAsync(CompletableFuture<PopConsumerContext> future,
+        String clientHost, String groupId, String topicId, int queueId, int batchSize, MessageFilter filter,
+        PopConsumerRecord.RetryType retryType) {
+
+        return future.thenCompose(result -> {
+
+            // pop request too much, should not add rest count here
+            if (isPopShouldStop(groupId, topicId, queueId)) {
+                return CompletableFuture.completedFuture(result);
+            }
+
+            // Current requests would calculate the total number of messages
+            // waiting to be filtered for new message arrival notifications in
+            // the long-polling service, need disregarding the backlog in order
+            // consumption scenario. If rest message num including the blocked
+            // queue accumulation would lead to frequent unnecessary wake-ups
+            // of long-polling requests, resulting unnecessary CPU usage.
+            // When client ack message, long-polling request would be notifications
+            // by AckMessageProcessor.ackOrderly() and message will not be delayed.
+            if (result.isFifo() && isFifoBlocked(result, groupId, topicId, queueId)) {
+                // should not add accumulation(max offset - consumer offset) here
+                return CompletableFuture.completedFuture(result);
+            }
+
+            int remain = batchSize - result.getMessageCount();
+            if (remain <= 0) {
+                result.addRestCount(this.getPendingFilterCount(groupId, topicId, queueId));
+                return CompletableFuture.completedFuture(result);
+            } else {
+                long consumeOffset = brokerController.getConsumerOffsetManager().queryPullOffset(groupId, topicId, queueId);
+                return getMessageAsync(clientHost, groupId, topicId, queueId, consumeOffset, remain, filter)
+                    .thenApply(getMessageResult -> addGetMessageResult(
+                        result, getMessageResult, topicId, queueId, retryType, consumeOffset));
+            }
+        });
+    }
+
+    public CompletableFuture<PopConsumerContext> popAsync(String clientHost, long popTime, long invisibleTime,
+        String groupId, String topicId, int queueId, int batchSize, boolean fifo, String attemptId,
+        MessageFilter filter) {
+
+        PopConsumerContext popConsumerContext =
+            new PopConsumerContext(clientHost, popTime, invisibleTime, groupId, fifo, attemptId);
+
+        TopicConfig topicConfig = brokerController.getTopicConfigManager().selectTopicConfig(topicId);
+        if (topicConfig == null || !consumerLockService.tryLock(groupId, topicId)) {
+            return CompletableFuture.completedFuture(popConsumerContext);
```

Comment:
Can this lock granularity be optimized to queue id level?  Don't see the need for a larger lock.

Reply:
In most scenarios, it is recommended to use pop broker(-1), and there is a unnessary cost of repeatedly locking and unlocking when there are a large number of queues